File: 11995 I Can Guess the Data Structure!/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
        
void append(short **structure, short *size, short data) {
    *structure = (short*)realloc(*structure, (*size + 1) * sizeof(short));
    (*structure)[(*size)++] = data;
}

void removeElement(short **structure, short *size, int index) {
    for(int i = index; i < *size - 1; ++i)
        (*structure)[i] = (*structure)[i + 1];
    --(*size);
}

int contain(short *structure, short size, short data) {
    for(int i = 0; i < size; ++i) {
        if(structure[i] == data)
            return i;
    }
    return -1;
}

int main() {
    int n;
    while(scanf("%d", &n) == 1) {
        short *stack = NULL, *queue = NULL, *priorityQueue = NULL;
        short stackSize = 0, queueSize = 0, priorityQueueSize = 0, max = 1;
        bool isStack = true, isQueue = true, isPriorityQueue = true;
        while(n--) {
            int command;
            short x;
            scanf("%d %hd", &command, &x);
            if(command == 1) {
                append(&stack, &stackSize, x);
                append(&queue, &queueSize, x);
                append(&priorityQueue, &priorityQueueSize, x);
                max = max > x ? max : x;
            } else if(command == 2) {
                int index = contain(priorityQueue, priorityQueueSize, x);
                if(index == -1)
                    isStack = isQueue = isPriorityQueue = false;
                else {
                    if(isStack) {
                        if(x == stack[stackSize - 1])
                            removeElement(&stack, &stackSize, stackSize - 1);
                        else
                            isStack = false;
                    }
                    if(isQueue) {
                        if(x == queue[0])
                            removeElement(&queue, &queueSize, 0);
                        else
                            isQueue = false;
                    }
                    removeElement(&priorityQueue, &priorityQueueSize, index);
                    if(x == max) {
                        max = 1;
                        for(int i = 0; i < priorityQueueSize; ++i)
                            max = max > priorityQueue[i] ? max : priorityQueue[i];
                    } else
                        isPriorityQueue = false;
                }
            }
        }
        if(isStack + isQueue + isPriorityQueue == 0)
			puts("impossible");
		else if(isStack + isQueue + isPriorityQueue >= 2)
			puts("not sure");
		else if(isStack)
			puts("stack");
		else if(isQueue)
			puts("queue");
		else if(isPriorityQueue)
			puts("priority queue");
		free(stack);
		free(queue);
		free(priorityQueue);
    }
    return 0;
}

File: 11995 I Can Guess the Data Structure!/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    StreamTokenizer st = new StreamTokenizer(System.in);
	    StringBuilder output = new StringBuilder();
	    while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            int n = (int)st.nval;
            short max = 1;
	        LinkedList<Short> stack = new LinkedList<>();
    		LinkedList<Short> queue = new LinkedList<>();
    		LinkedList<Short> priorityQueue = new LinkedList<>();
    		boolean isStack = true, isQueue = true, isPriorityQueue = true;
    		while(n-- > 0) {
    		    st.nextToken();
    			int command = (int)st.nval;
    			st.nextToken();
    			short x = (short)st.nval;
    			if(command == 1){
    				stack.add(x);
    				queue.add(x);
    				priorityQueue.add(x);
    				max = (short)Math.max(max, x);
    			} else if(command == 2) {
					if(!priorityQueue.contains(x))
    					isStack = isQueue = isPriorityQueue = false;
    				else {
    					if(isStack && x != stack.removeLast())
    						isStack = false;
						if(isQueue && x != queue.removeFirst())
    						isQueue = false;
    					priorityQueue.remove(Short.valueOf(x));
    					if(x == max) {
							max = 1;
    						for(Short s : priorityQueue)
    							max = (short)Math.max(max, s);
    					} else
							isPriorityQueue = false;	
    				}
    			}
    		}
    		if(!isStack && !isQueue && !isPriorityQueue)
    			output.append("impossible\n");
			else if((isStack && isQueue) || (isQueue && isPriorityQueue) || (isStack && isPriorityQueue))
    			output.append("not sure\n");
    		else if(isStack)
				output.append("stack\n");
    		else if(isQueue)
    			output.append("queue\n");
			else if(isPriorityQueue)
    			output.append("priority queue\n");
	    }
	    System.out.print(output);
	}
}
